### ISSUE TITLE
F# option (FSharpOption<T>) fails to deserialize

### ISSUE THREAD

**Issue by ca-ta** [dotnet/orleans#1346]
Adding an interface that uses `FSharpOption<T>` results in compile-time serializers being generated. This breaks the deserialization of the F# `option` type.

**Comment by ca-ta** [comment]
Repro available in  #1347

**Comment by ca-ta** [comment]
@ReubenBond - I was wondering if you had a chance to give this and/or #1349 a look

**Comment by ca-ta** [comment]
Repro PR on hold - waiting for #1369 to be merged.

**Comment by ca-ta** [comment]
@ReubenBond - It would be great if you could give this a look. Repro available in #1417

I have tried to fix the issue but failed - a `System.Security.VerificationException` get triggered at deserialization time, with the message `Operation could destabilize the runtime.` The stack trace is:
```
Test Name:	UnitTests.Serialization.SerializationTestsFsharpTypes.SerializationTests_FSharp_IntOption_Some
Test FullName:	UnitTests.Serialization.SerializationTestsFsharpTypes.SerializationTests_FSharp_IntOption_Some
Test Source:	C:\code\ca-ta\orleans\test\Tester\SerializationTests\SerializationTests.FSharpTypes.cs : line 28
Test Outcome:	Failed
Test Duration:	0:00:03.61

Result StackTrace:	
at valueSet(FSharpOption`1 , Int32 )
   at UnitTests.GrainInterfaces.OrleansCodeGenMicrosoft_FSharp_Core_FSharpOptionSerializer`1.Deserializer(Type expected, BinaryTokenStreamReader stream) in C:\code\ca-ta\orleans\src\TestGrainInterfaces\Properties\orleans.codegen.cs:line 7760
   at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, BinaryTokenStreamReader stream) in C:\code\ca-ta\orleans\src\Orleans\Serialization\SerializationManager.cs:line 1525
   at Orleans.Serialization.SerializationManager.DeserializeInner[T](BinaryTokenStreamReader stream) in C:\code\ca-ta\orleans\src\Orleans\Serialization\SerializationManager.cs:line 1435
   at Orleans.Serialization.SerializationManager.DeserializeFromByteArray[T](Byte[] data) in C:\code\ca-ta\orleans\src\Orleans\Serialization\SerializationManager.cs:line 1709
   at Orleans.Serialization.SerializationManager.RoundTripSerializationForTesting[T](T source) in C:\code\ca-ta\orleans\src\Orleans\Serialization\SerializationManager.cs:line 2203
   at UnitTests.Serialization.SerializationTestsFsharpTypes.RoundtripSerializationTest[T](T input) in C:\code\ca-ta\orleans\test\Tester\SerializationTests\SerializationTests.FSharpTypes.cs:line 22
   at UnitTests.Serialization.SerializationTestsFsharpTypes.SerializationTests_FSharp_IntOption_Some() in C:\code\ca-ta\orleans\test\Tester\SerializationTests\SerializationTests.FSharpTypes.cs:line 29
Result Message:	System.Security.VerificationException : Operation could destabilize the runtime.
```
I've run `peverify` on the grain interfaces assembly (the one containing the generated serializers), but no errors were found. Also, I've debugged the tests and the types seem correct... and yet the tests fail.

I have also experimented with temporarily suppressing interface code gen (by commenting the Orleans targets in `TestGraininterfaces.csproj`) and manually commenting out the generated `OrleansCodeGenMicrosoft_FSharp_Core_FSharpOptionSerializer<T>`. While this change **did** allow the tests to pass, attempting the same approach in our system did not work (still getting the exception in `valueSet`) - so I am quite stumped.

Needless to say, this issue prevents us (and I suspect others using F#) from upgrading to any Orleans version newer than 1.0.9. 
Any suggestions on how to attack this issue (as well as any workarounds/hacks that may come to mind) are very welcome.

**Comment by ca-ta** [comment]
@RreubenBond - any feedback on this one? Note that #1424  and #1422 may be related?

**Comment by ReubenBond** [comment]
Fixed in #1504, specifically [this commit](https://github.com/ReubenBond/orleans/commit/ba77700c6d65cb1366b2f964f6286dc339e3b178). There was a security/trust issue which was remedied by emitting the field acessor `DynamicMethod`s into the same module as the field's type instead of into the declaring type's module. I'm not completely certain that there will never be a case where the declaring type has a higher security requirement than the field - it seems plausible. In that case we will have to investigate other options such as raising the trust level of Orleans.dll. Hopefully this all goes away in .NET Core.

**Comment by ca-ta** [comment]
@ReubenBond  - thanks, and thanks for the commit pointer - I'll update #1417 shortly.

**Comment by ReubenBond** [comment]
Fixed by #1504
